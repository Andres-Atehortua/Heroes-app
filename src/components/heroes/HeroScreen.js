import { useMemo } from 'react';
import { Redirect, useHistory, useParams } from 'react-router-dom';
import { getHeroById } from '../../selectors/getHeroById';

const HeroScreen = () => {
  const { heroId } = useParams();

  const hero = useMemo(() => getHeroById(heroId), [heroId]);
  const history = useHistory();

  if (!hero) {
    return <Redirect to='/' />;
  }

  const handleReturn = () => {
    history.push(`/${heroId.split('-')[0]}`);
  };

  const {
    superhero,
    publisher,
    alter_ego,
    first_appearance,
    characters,
  } = hero;

  return (
    <div className='row mt-5'>
      <div className='col-4'>
        <img
          className='img-thumbnail animate__animated animate__fadeInLeft'
          src={`../assets/${heroId}.jpg`}
          alt={superhero}
        />
      </div>
      <div className='col-8 animate__animated animate__fadeInUp'>
        <h3>{superhero}</h3>
        <ul className='list-group list-group-flush'>
          <li className='list-group-item'>
            <b>Alter ego:</b> {alter_ego}
          </li>
          <li className='list-group-item'>
            <b>Publisher:</b> {publisher}
          </li>
          <li className='list-group-item'>
            <b>First appearance:</b> {first_appearance}
          </li>
        </ul>
        <h5>Characters</h5>
        <p>{characters}</p>

        <button onClick={handleReturn} className='btn btn-outline-info'>
          Return
        </button>
      </div>
    </div>
  );
};

export default HeroScreen;
