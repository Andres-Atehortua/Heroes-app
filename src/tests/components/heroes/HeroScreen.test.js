import { mount } from 'enzyme';
import { Router, MemoryRouter, Route } from 'react-router-dom';

import HeroScreen from '../../../components/heroes/HeroScreen';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
}));

describe('Testing component <HeroScreen />', () => {
  const historyMock = {
    pathname: {},
    push: jest.fn(),
    listen: jest.fn(),
    location: {},
    replace: jest.fn(),
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should show <Redirect /> if arguments are not passed', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero']}>
        <HeroScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Redirect').exists()).toBe(true);
  });

  test('should show a hero if param exist', () => {
    const wrapper = mount(
      <Router history={historyMock}>
        {/* Con lo de abajo simulamos par√°metros */}
        <MemoryRouter initialEntries={['/hero/marvel-spider']}>
          <Route path='/hero/:heroId'>
            <HeroScreen />
          </Route>
        </MemoryRouter>
      </Router>
    );

    expect(wrapper.find('.row').exists()).toBe(true);
  });

  test('should go to the back section when press back', () => {
    // const history = {
    //   push: jest.fn(),
    //   listen: jest.fn(),
    //   replace: jest.fn(),
    //   location: {},
    // };
    // const wrapper = mount(
    //   // <Router history={history}>
    //   <MemoryRouter initialEntries={['/hero/marvel-spider']}>
    //     <Route path='/hero/:heroId'>
    //       <HeroScreen />
    //     </Route>
    //   </MemoryRouter>
    //   // </Router>
    // );
    // wrapper.find('button').simulate('click');
    // expect(history.push).toHaveBeenCalled();
  });
});
