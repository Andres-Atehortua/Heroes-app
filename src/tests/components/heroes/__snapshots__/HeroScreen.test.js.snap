// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing component <HeroScreen /> should show <Redirect /> if arguments are not passed 1`] = `""`;
